### ISSUE TITLE
Index Checker Limitation: Type refinement for entire expressions

### ISSUE THREAD

**Issue by as56** [kelloggm/checker-framework#94]
consider the code:
```
import org.checkerframework.checker.index.qual.*;
public class Limitation {
	
	public static void m1() {
		int[] a = {1,2,3,4,5};
		@IndexFor("a") int i = 4;
		@IndexFor("a") int j = 4;
		if(2*j <= i) {
			@IndexFor("a") int k = 2*j;//index checker warning
		}	
	}

}
```
Notice in the above code if 2*j is less or equal to i which is an @IndexFor("a") type. Therefore, the whole expression 2*j is also of type @IndexFor("a"). But the checker issues an upperbound unknown warning inside the if statement. Of course, this cannot be generalized to all expressions, but the checker ideally should not warn for any expression that produces a non-negative result. And if the expression in question produces a negative result, the lowerbound sub-checker should issue the appropriate warning.

Checker Warning:
Limitation.java:14: error: [assignment.type.incompatible] incompatible types in assignment.
			@IndexFor("a") int k = 2*j;		
			                        ^
  found   : @UpperBoundUnknown int
  required: @LTLengthOf("a") int
1 error

**Comment by kelloggm** [comment]
Duplicate with #52. This would be very valuable to add, I agree, but we haven't seen it come up in practice. Does this come up in one of the case studies you're working on?

**Comment by as56** [comment]
Hi Martin, this does come up a few times in my case studies, and I have seen code written like this many times in practice.

**Comment by mernst** [comment]
This doesn't look to me like it is related to #52.

**Comment by kelloggm** [comment]
You're right. For some reason (probably the title) I thought that that issue referred to a different problem and didn't check.

**Comment by kelloggm** [comment]
So the precise problem here is that expressions in general can't benefit from transfer functions, because their constituent parts could be changed. The offsets work that suzanne is doing solves this problem for addition and subtraction, but in some sense both of those are special cases of this more general problem. 

If we wanted to do this in general, we'd need to assign a type to each subexpression. I don't think the CF does that. An alternative way for the IC to do this would be to check which expressions are used as indexes, and then push those back into local variables if possible.

**Comment by mernst** [comment]
The store maps from (somewhat arbitrary) expressions to types.  The Java expressions that can be mapped from are described at https://checkerframework.org/manual/#java-expressions-as-arguments .  That section says:

> Limitations: The following Java expressions may not currently be written:
> Mathematical operators (plus, minus, division, ...).

Lifting the limitation on mathematical operators would solve this issue.

**Comment by kelloggm** [comment]
I've been looking into lifting this restriction. I'm making this note for my own benefit.

Adding support for mathematical operators appears to be possible, but requires a non-trivial amount of work. In particular, it requires (at least) a new parsing method in FlowExpressionParseUtil and a new kind of Receiver in FlowExpressions (which the parsing method will want to return). FlowExpressions.internalReprOf and associated methods will also need to be modified. I think it would be wise to discuss this with @smillst and/or @mernst before I going rushing in blindly, since the code I would need to modify is quite integral to how the CF works.
